go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,482 @@
+* No need to pass a handler, that is just part of the last item in the interceptor chain (i.e. when calling next)
+
+### WorkflowOutboundInterceptor
+
+We can leave everything the way it is because, while it doesn't make the interceptor parameters future-proof with struct
+wrappers, it mimics the actual calls that aren't going to have backwards-compatible alterations.
+
+### WorkflowOutboundInterceptor.GetInfo
+
+Current:
+
+```go
+GetWorkflowInfo(ctx Context) *WorkflowInfo
+```
+
+Proposal:
+
+```go
+GetInfo(ctx Context) *WorkflowInfo
+```
+
+Why: All other calls match `workflow` package, why not this?
+
+Note: We need to make sure we document that this returns *workflow.Info since internal names can be confusing.
+
+### WorkerInterceptor.InterceptActivity
+
+Proposal:
+
+```go
+InterceptActivity(ctx context.Context, next ActivityInboundInterceptor) ActivityInboundInterceptor
+```
+
+Why: Mimics `InterceptWorkflow` but for activities.
+
+### ActivityInboundInterceptor
+
+Proposal:
+
+```go
+type ActivityInput struct {
+  Args []interface{}
+}
+
+type ActivityInboundInterceptor interface {
+  Init(outbound ActivityOutboundInterceptor) error
+
+  ExecuteActivity(ctx context.Context, input *ActivityInput) (interface{}, error)
+
+  mustEmbedActivityInboundInterceptorBase()
+}
+```
+
+Also a `ActivityInboundInterceptorBase` implementation with expected delegation.
+
+Why: Mimics `WorkflowInboundInterceptor` but for activities.
+
+### ActivityOutboundInterceptor
+
+Proposal:
+
+```go
+type ActivityOutboundInterceptor interface {
+  GetInfo(ctx context.Context) ActivityInfo
+  GetLogger(ctx context.Context) log.Logger
+  GetMetricsScope(ctx context.Context) tally.Scope
+  RecordHeartbeat(ctx context.Context, details ...interface{})
+  HasHeartbeatDetails(ctx context.Context) bool
+  GetHeartbeatDetails(ctx context.Context, d ...interface{}) error
+  GetWorkerStopChannel(ctx context.Context) <-chan struct{}
+
+  mustEmbedActivityOutboundInterceptorBase()
+}
+```
+
+Also a `ActivityOutboundInterceptorBase` implementation with expected delegation.
+
+Why:
+
+* Mimics `WorkflowOutboundInterceptor` but for activities.
+
+For discussion: `GetInfo` returns expensive copies because `activity.GetInfo` does. Is this ok? This is inconsistent
+with workflow info which is a pointer.
+
+### ClientInterceptor
+
+Proposal:
+
+```go
+type ClientInterceptor interface {
+  InterceptClient(next ClientOutboundInterceptor) ClientOutboundInterceptor
+
+  mustEmbedClientInterceptorBase()
+}
+```
+
+Why:
+
+* This kinda mimics `WorkerInterceptor` calls, but note it only provides an outbound interceptor because all client
+  calls are outbound by their nature so there is no inbound equivalent
+
+### ClientOutboundInterceptor
+
+Proposal:
+
+```go
+type ClientOutboundInterceptor interface {
+  ExecuteWorkflow(ctx context.Context, options StartWorkflowOptions, workflow interface{}, args ...interface{}) (WorkflowRun, error)
```

Comment:
I might as well go ahead and revert https://github.com/temporalio/proposals/pull/45/commits/14f4e43c7930b8bd4e1021ea4bcb71def87c8a2d since our params will now not be matching the client call necessarily

Reply:
I have updated the proposal to reflect this change in thinking.